product_supplierinfo_group/models/product_supplierinfo.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2020 Akretion France (http://www.akretion.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+
+from odoo import api, fields, models
+
+# format: field_from_supplierinfo:field_from_group
+MAPPING_RELATED = {
+    "company_id": "company_id",
+    "product_tmpl_id": "product_tmpl_id",
+    "name": "partner_id",
+    "product_id": "product_id",
+    "product_name": "product_name",
+    "product_code": "product_code",
+    "sequence": "sequence",
+}
+
+MAPPING_MATCH_GROUP = {
+    "company_id": "company_id",
+    "product_tmpl_id": "product_tmpl_id",
+    "name": "partner_id",
+    "product_id": "product_id",
+    "product_name": "product_name",
+    "product_code": "product_code",
+}
+
+_logger = logging.getLogger(__name__)
+
+
+class ProductSupplierinfo(models.Model):
+    _inherit = "product.supplierinfo"
+
+    group_id = fields.Many2one("product.supplierinfo.group", required=True)
+    company_id = fields.Many2one(related="group_id.company_id", store=True)
+    product_tmpl_id = fields.Many2one(related="group_id.product_tmpl_id", store=True)
+    name = fields.Many2one(related="group_id.partner_id", store=True, required=False)
+    product_id = fields.Many2one(related="group_id.product_id", store=True)
+    product_name = fields.Char(related="group_id.product_name", store=True)
+    product_code = fields.Char(related="group_id.product_code", store=True)
+    sequence = fields.Integer(related="group_id.sequence", store=True)
+
+    _sql_constraints = [
+        (
+            "uniq_price_per_qty",
+            "unique(group_id, min_qty, date_start, date_end)",
+            "You can not have a two price for the same qty",
+        )
+    ]
+
+    def _fields_for_group_match(self):
+        return MAPPING_MATCH_GROUP
+
+    def _get_or_create_group(self, vals):
+        field_mapping = self._fields_for_group_match().items()
+        group = self.env["product.supplierinfo.group"].search(
+            [
+                (field_group, "=", vals.get(field_supplierinfo))
+                for field_supplierinfo, field_group in field_mapping
+            ]
+        )
+        if not group:
+            group = self.env["product.supplierinfo.group"].create(
+                {
+                    field_group: vals.get(field_supplierinfo)
+                    for field_supplierinfo, field_group in field_mapping
+                }
+            )
+        return group
+
+    @api.model_create_multi
+    def create(self, list_vals):
+        for vals in list_vals:
+            if not vals.get("group_id"):
+                vals["group_id"] = self._get_or_create_group(vals).id
+                # remove useless key
+                for key in MAPPING_RELATED.keys():
```

Comment:
Would it be possible to call a method just as `_fields_for_group_match` but to return `MAPPING_RELATED`?

From what I understand, MAPPING_MATCH_GROUP need to be outside the class so it can be used in the pre init hook, but it is overiddable because of `_fields_for_group_match`

I guess we should have the same thing for `MAPPING_RELATED`. So the module will be fully overidable in case of additional custom fields on supplierinfo, that should be in the group part.